node/pkg/watchers/sui/watcher.go

```diff
@@ -0,0 +1,355 @@
+	}
+
+	logger.Info("Subscribing using", zap.String("filter", s))
+
+	if _, err := ws.Write([]byte(s)); err != nil {
+		logger.Error(fmt.Sprintf("write: %s", err.Error()))
+		return err
+	}
+
+	timer := time.NewTicker(time.Second * 1)
+	defer timer.Stop()
+
+	supervisor.Signal(ctx, supervisor.SignalHealthy)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case r := <-e.obsvReqC:
+			if vaa.ChainID(r.ChainId) != vaa.ChainIDSui {
+				panic("invalid chain ID")
+			}
+
+			id := base64.StdEncoding.EncodeToString(r.TxHash)
+
+			logger.Info("obsv request", zap.String("TxHash", string(id)))
+
+			buf := fmt.Sprintf(`{"jsonrpc":"2.0", "id": 1, "method": "sui_getEvents", "params": [{"Transaction": "%s"}, null, 10, true]}`, id)
+
+			resp, err := http.Post(e.suiRPC, "application/json", strings.NewReader(buf))
+			if err != nil {
+				logger.Error(e.suiRPC, zap.Error(err))
+				p2p.DefaultRegistry.AddErrorCount(vaa.ChainIDSui, 1)
+				continue
+			}
+
+			body, err := io.ReadAll(resp.Body)
+			if err != nil {
+				logger.Error(e.suiRPC, zap.Error(err))
+				p2p.DefaultRegistry.AddErrorCount(vaa.ChainIDSui, 1)
+				continue
+
+			}
+			logger.Info(string(body))
+
+			if !gjson.Valid(string(body)) {
+				logger.Error("InvalidJson: " + string(body))
+				p2p.DefaultRegistry.AddErrorCount(vaa.ChainIDAptos, 1)
+				break
+			}
+
+			outcomes := gjson.ParseBytes(body).Get("result.data")
+			if !outcomes.Exists() {
+				logger.Error("InvalidJson: " + string(body))
+				p2p.DefaultRegistry.AddErrorCount(vaa.ChainIDAptos, 1)
+				break
+			}
+
+			for _, chunk := range outcomes.Array() {
+				err := e.inspectBody(logger, chunk)
+				if err != nil {
+					logger.Error(e.suiRPC, zap.Error(err))
+				}
+			}
+
+		case <-timer.C:
+			for {
+				var msg = make([]byte, 10000)
+				err := ws.SetReadDeadline(time.Now().Local().Add(100_000_000))
+				if err != nil {
+					return err
+				}
+
+				if n, err := ws.Read(msg); err != nil {
+					if err.Error() == "EOF" {
+						return err
+					}
+					break
+				} else {
+					parsedMsg := gjson.ParseBytes(msg)
+					if !parsedMsg.Exists() {
+						logger.Error("error", zap.String("body", string(msg[:n])), zap.Uint64("len", uint64(n)))
+						continue
+					}
+					logger.Info("receive", zap.String("body", string(msg[:n])), zap.Uint64("len", uint64(n)))
+					result := parsedMsg.Get("params.result")
+					if result.Exists() {
+						err := e.inspectBody(logger, result)
+						if err != nil {
+							logger.Error(fmt.Sprintf("inspectBody: %s", err.Error()))
+						}
+						continue
+					}
+					errVal := parsedMsg.Get("error.message")
+					if errVal.Exists() {
+						return errors.New(errVal.String())
+					}
+
+					id := parsedMsg.Get("id")
+					if id.Exists() {
+						if id.Int() == e.subId {
+							e.subscribed = true
+						}
+						continue
+					}
```

Comment:
Rather than doing this within the loop can we handle the subscription response beforehand?  IIUC, we should receive exactly one message like this right at the beginning.

Reply:
We could break this out into its own loop but the response does not happen instantly but occures at some point on the websocket.  It would not make the code any clearer